library/java/org/chromium/net/impl/CronetBidirectionalState.java

```diff
@@ -0,0 +1,431 @@
+    int WRITING = 1 << 5;               // One RequestBody's Buffer is being sent on the wire.
+    int READING = 1 << 6;               // One ResponseBody's Buffer is being read from the wire.
+    int HEADERS_SENT = 1 << 7;          // EM's "sendHeaders" method has been invoked.
+    int CANCELLING = 1 << 8;            // EM's "cancel" method has been invoked.
+    int USER_CANCELLED = 1 << 9;        // The cancel operation was initiated by the User.
+    int FAILED = 1 << 10;               // An fatal failure has been encountered.
+    int ON_HEADER_RECEIVED = 1 << 11;   // EM's "onHeaders" callback has been invoked.
+    int ON_COMPLETE_RECEIVED = 1 << 12; // EM's "onComplete" callback has been invoked.
+    int READ_DONE = 1 << 13;            // User won't receive more read callbacks.
+    int WRITE_DONE = 1 << 14;           // User won't receive more write callbacks. Maybe never had.
+    int DONE = 1 << 15;                 // Terminal state. Can be successful or otherwise.
+    int TERMINATING_STATES = CANCELLING | FAILED | DONE; // Hold your breath and count to ten.
+  }
+
+  private final AtomicInteger mState = new AtomicInteger(State.NOT_STARTED);
+
+  /**
+   * Returns true if the final state has been reached. At this point the EM Stream has been
+   * destroyed.
+   */
+  boolean isDone() { return (mState.get() & State.DONE) != 0; }
+
+  /**
+   * Returns true if a terminating state has been reached. Terminating does not necessarily means
+   * that the DONE state has been reached. When the DONE bit is not set, it means that we are not
+   * ready yet to inform the user about the failure, as the EM as not yet destroyed the Stream. In
+   * other words, EM has not yet invoked a terminal callback (onError, onCancel, onComplete).
+   */
+  boolean isTerminating() { return (mState.get() & State.TERMINATING_STATES) != 0; }
+
+  /**
+   * Returns the reason why the request finished. Can only be invoked if {@link #isDone} returns
+   * true.
+   *
+   * @return one of {@link RequestFinishedInfo#SUCCEEDED}, {@link RequestFinishedInfo#FAILED}, or
+   *     {@link RequestFinishedInfo#CANCELED}
+   */
+  @FinishedReason
+  int getFinishedReason() {
+    assert isDone();
+    @State int finalState = mState.get();
+    if ((finalState & State.FAILED) != 0) {
+      return RequestFinishedInfo.FAILED;
+    }
+    if ((finalState & State.USER_CANCELLED) != 0) {
+      return RequestFinishedInfo.CANCELED;
+    }
+    return RequestFinishedInfo.SUCCEEDED;
+  }
+
+  /**
+   * Establishes what is the next action by taking in account the current global state, and the
+   * provided {@link Event}. This method has one important side effect: the resulting global state
+   * is saved through an Atomic operation. For few cases, this method will throw when the state is
+   * not compatible with the event.
+   */
+  @NextAction int nextAction(@Event final int event) { // "final" just to avoid dumb mistakes.
+    while (true) {
```

Comment:
I don't think I understand why this is in a loop. The method is quite big so maybe I'm not able to hold it all in memory. Would it be possible to extract the body of this `while (true) {` out into a stand-alone method which does a single iteration of the loop?

Reply:
Since Even only has a few values, we could encode both nextAction and nextState into a single long (and probably even an int) and just return that, right? That being said, thanks for extracting out the isAlreadyFinalState() method. That helps!

Reply:
Discussed in person. Conclusion: it is doable to to have a state for "read" and another "state" for write. However, there is some added trickiness by following that path: can foster some race conditions. Nothing that can't be circumvented, but this is not trivial. Status quo for the moment.